presto-hive/src/main/java/com/facebook/presto/hive/s3select/S3SelectRecordCursorProvider.java

```diff
@@ -97,4 +110,65 @@ public Optional<RecordCursor> createRecordCursor(
         // unsupported serdes
         return Optional.empty();
     }
+
+    private static boolean hasFilters(
+            Properties schema,
+            TupleDomain<HiveColumnHandle> effectivePredicate,
+            List<HiveColumnHandle> projectedColumns)
+    {
+        // When there are no effective predicates and the projected columns and column types are identical to the schema,
+        // it means we get all the data out of S3. We can use S3 GetObject instead of S3 SelectObjectContent in these cases.
+        if (effectivePredicate.isAll()) {
+            return !isSchemaEquivalent(projectedColumns, schema);
+        }
+        return true;
+    }
+
+    private static boolean isSchemaEquivalent(List<HiveColumnHandle> columns, Properties schema)
+    {
+        Set<String> projectedColumnNames = getColumnProperty(columns, HiveColumnHandle::getName);
+        Set<String> projectedColumnTypes = getColumnProperty(columns, column -> column.getHiveType().getTypeInfo().getTypeName());
+        return areColumnsEquivalent(projectedColumnNames, schema) && areColumnTypesEquivalent(projectedColumnTypes, schema);
+    }
+
+    private static boolean areColumnsEquivalent(Set<String> projectedColumnNames, Properties schema)
+    {
+        Set<String> columnNames;
+        String columnNameProperty = schema.getProperty(LIST_COLUMNS);
+        if (columnNameProperty.length() == 0) {
+            columnNames = ImmutableSet.of();
+        }
+        else {
+            String columnNameDelimiter = (String) schema.getOrDefault(COLUMN_NAME_DELIMITER, ",");
+            columnNames = Arrays.stream(columnNameProperty.split(columnNameDelimiter))
```

Comment:
Seems like this could be `ImmutableSet.copyOf(columnNameProperty.split(columnNameDelimiiter))`

Reply:
Thanks! Nice catch!